File: heap/lastStoneWeight.py
# https://leetcode.com/problems/last-stone-weight/KW

import heapq

class Solution(object):
    def lastStoneWeight(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
        while len(stones)>=2:
            stones.sort()
            y = stones.pop()
            x = stones.pop()
            
            if y-x :stones.append(y-x)
        
        if stones:
            return stones[-1]
        else:
            return 0
        
    
    def lastStoneWeight(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
        h = [-ele for ele in stones]
        heapq.heapify(h)
        
        while len(h)>1:
            heapq.heappush(h, (heapq.heappop(h)- heapq.heappop(h)))
            
        return -h[0]

File: heap/lastStoneWeight.java
//Leetcode = https://leetcode.com/problems/last-stone-weight/
import java.util.*;

class lastStoneWeight
 {
    public int lastStoneWeight(int[] stones) {
        
        PriorityQueue<Integer> pq  = new PriorityQueue<>(Collections.reverseOrder());
        
        for(int i: stones){
            pq.add(i);
        }
        
        while(pq.size()>1){
            int y = pq.poll();
            int x = pq.poll();
            if(y != x){
                pq.add(y-x);
            }
        }
        return pq.size()==0?0:pq.poll();
    }
}